src/fm_classifier.jl

```diff
@@ -0,0 +1,84 @@
+type FMClassifier <: FMPredictor
+    k0::Bool
+    k1::Bool
+
+    # model parameters
+    w0::FMFloat
+    w::Array{Float64,1}
+    V::Array{Float64,2}
+
+    # regularization
+    reg0::FMFloat
+    regw::FMFloat
+    regv::Array{FMFloat}
+
+    initMean::FMFloat
+    initStdev::FMFloat
+
+    num_attribute::FMInt
+    num_factor::FMInt
+end
+
+function fmPredictInstance!(fm::FMClassifier, idx::Array{Int64,1}, x::Array{FMFloat,1}, fSum::Array{FMFloat}, sum_sqr::Array{FMFloat})
+   fill!(fSum, .0)
+   fill!(sum_sqr, .0)
+   result = .0
+   if(fm.k0)
+       result += fm.w0
+   end
+   if(fm.k1)
+       for i in 1:length(idx)
+            result += fm.w[idx[i]] * x[i]
+       end
+   end
+   for f in 1:fm.num_factor
+       for i in 1:length(idx)
+           d = fm.V[f,idx[i]] * x[i]
+           fSum[f] += d
+           sum_sqr[f] += d*d
+       end
+       result += 0.5 * (fSum[f]*fSum[f] - sum_sqr[f])
+   end
+    # squash result between [0, 1]
+    1.0 / (1.0 + exp(-result))
+end
+
+fmLoss(::Type{FMClassifier}, yhat::FMFloat, y::FMFloat) = (1 - sign(yhat * y)) / 2
+fmLossGradient(::Type{FMClassifier}, yhat::FMFloat, y::FMFloat) = -y * (1.0 - 1.0 / (1.0 + exp(-y * yhat)))
```

Comment:
Should probably note that this is the gradient with respect to the prediction (not the gradient with respect to the parameters of the prediction).